### PR TITLE
Feature/te 388 pretty print

### PR DESCRIPTION
**Pull request by gunther-bachmann**

### CODE REVIEW

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/AbstractTclFormatterTest.xtend [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.testeditor.tcl.dsl.tests.formatter
+
+import javax.inject.Inject
+import org.testeditor.tcl.dsl.tests.AbstractTclTest
+import com.google.inject.Provider
+import org.eclipse.xtext.junit4.formatter.FormatterTestRequest
+import org.eclipse.xtext.junit4.formatter.FormatterTester
+
+class AbstractTclFormatterTest extends AbstractTclTest {
+	@Inject Provider<FormatterTestRequest> formatterRequestProvider
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/AbstractTclFormatterTest.xtend]: unused

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/AbstractTclFormatterTest.xtend [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+	@Inject protected FormatterTester formatterTester
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/AbstractTclFormatterTest.xtend]: I would register this as an extension provider, i.e. `protected extension FormatterTester` and leave out the variable name in subclasses. This has the advantage that we could override `assertFormatted` in the future without touching all subclasses. Had this problem in another project (i.e. adjust the formatting request).

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/AbstractTclFormatterTest.xtend [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package org.testeditor.tcl.dsl.tests.formatter
+
+import javax.inject.Inject
+import org.testeditor.tcl.dsl.tests.AbstractTclTest
+import com.google.inject.Provider
+import org.eclipse.xtext.junit4.formatter.FormatterTestRequest
+import org.eclipse.xtext.junit4.formatter.FormatterTester
+
+class AbstractTclFormatterTest extends AbstractTclTest {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/AbstractTclFormatterTest.xtend]: Adding the following utility method is very useful for subclasses:
```
	/**
	 * Converts the passed String into a single line and replaces all whitespace sequences
	 * by a double space.
	 */
	protected def String toSingleLine(CharSequence input) {
		return input.toString.split(lineSeparator).join(' ').replaceAll('\\s+', '  ')
	}
```

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/TclModelFormatterTest.xtend [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+				
+				import a.b.c
+				import c.d.e
+				
+				require freq, breq
+				
+				# testCase
+				
+				* specification
+				Component: some
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/TclModelFormatterTest.xtend]: `Component` and everything below the specification should be indented.

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/TclModelFormatterTest.xtend [hunk 5]
```diff
@@ -0,0 +1,75 @@
+package org.testeditor.tcl.dsl.tests.formatter
+
+import org.junit.Test
+
+class TclModelFormatterTest extends AbstractTclFormatterTest {
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/TclModelFormatterTest.xtend]: I think there's too much going on in the two test methods. I would rather split them and handle the various constructs individually.

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/formatting2/TclFormatter.xtend [hunk 6]
```diff
@@ -13,14 +13,39 @@
...
-		
+class TclFormatter extends TmlFormatter {
+
+	def dispatch void format(TclModel tclModel, extension IFormattableDocument document) {
+		tclModel.regionFor.feature(TCL_MODEL__PACKAGE).append[newLines = 2]
+		// import section is not formatted (yet), should be done by organize imports
+		tclModel.regionFor.keyword("require").prepend[newLines = 2]
+		tclModel.regionFor.keyword(",").prepend[noSpace]
+		tclModel.environmentVariableReferences.forEach[prepend[oneSpace]]
+		tclModel.test.format
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/formatting2/TclFormatter.xtend]: You should also pass the `document` when calling `format` on elements.

File: tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/formatting2/TclFormatter.xtend [hunk 7]
```diff
@@ -13,14 +13,39 @@
...
+	}
+
+	def dispatch void format(TestCase testCase, extension IFormattableDocument document) {
+		testCase.regionFor.keyword("#").prepend[newLines = 2].append[oneSpace]
+		testCase.regionFor.feature(TEST_CASE__NAME).append[newLines = 2]
```
Comment (franzbecker) [inline tcl/org.testeditor.tcl.dsl/src/org/testeditor/tcl/dsl/formatting2/TclFormatter.xtend]: What aboud `implements`?

File: tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/TclModelFormatterTest.xtend (hunk 5 above)
Comment (gunther-bachmann) [inline tcl/org.testeditor.tcl.dsl.tests/src/org/testeditor/tcl/dsl/tests/formatter/TclModelFormatterTest.xtend]: splitted into two, several constructs are checked by the respective tml ...FormatterTest

**Pull request merged**